### ISSUE TITLE
Improving Altcoin support

### ISSUE THREAD

**Issue by NicolasDorier** [MetacoSA/NBitcoin#371]
So @Kukks came with an idea which seem a viable path for increasing altcoin support in NBitcoin.

# Current situation 

Current altcoin support is possible only if the transaction, block, blockheader, and way of calculating the block headerhash are similar to bitcoin. (like Litecoin, Viacoin, or BCash)
This is done by building a `Network` through the `NetworkBuilder` class.
One can specify a way to calculate the `PoWHash` algorithm as well so it is possible to check PoW on altcoins not using SHA256 as their PoW algorithm.

# Proposed change

In order to not break any existing code, @Kukks proposed a solution which might work and involve minimal work.

For block header for example, the Network class would have a `BlockHeaderFactory` instance, responsible for parsing a Block Header from bytes.
Non-Bitcoin block header would be subclass of block header, overriding `GetHash`, `GetPowHash` and  `ReadWrite`.
A block header instance would then know exactly, without any outside information: 

* How to calculate its hash, 
* How to calculate its PoW Hash, 
* How to serialize itself,

# Consequence

It means that in order to correctly `deserialize` a blockheader we would need to have a Network instance at hand. And luckily, this should be already the case. (Node class is an example)

Place deserializing a block header and not having Network/BlockHeaderFactory at hand should assume Bitcoin format and marked as issues in github for later refactoring.

# Breaking change

The `GetPoWHash` of `BlockHeader` currently takes a `Network` parameter. This is not needed anymore as the BlockHeader would get the logic for `GetPowHash` calculation inside itself.
We can just keep the current `GetPowHash(Network)`  and mark as obsolete.
Then having the real `GetPowHash()` without Network parameter

# Note

We would use the following plan for `BlockHeader/BlockHeaderFactory`, `Block/BlockFactory`, `Transaction/TransactionFactory`.

I hope we can get one separate PR for each of them, starting by `BlockHeader/BlockHeaderFactory`.

@Kukks @knocte @dangershony @bokobza @Aprogiena @nopara73

**Comment by nopara73** [comment]
Concept ACK. I think it's a great solution. Altcoiners may need to go for the extra mile to understand it, but as far as I understand it only minimally [messes with the Bitcoin user space](https://lkml.org/lkml/2012/12/23/75) (IMO `GetPowHash` and `BitcoinStream` are not often used), which is the most important.

**Comment by Aprogiena** [comment]
We did have a similar discussion in Stratis, but I wasn't involved in that that much. I think @dangershony can talk more about how we solved that. We certainly started the transition, not sure if we finished it at this point.

**Comment by Kukks** [comment]
You polished up the idea really well @NicolasDorier  👍 
Perhaps we can leave the `GetPowHash` there marked as Obsolete in case it breaks for other devs?

**Comment by NicolasDorier** [comment]
yes, I think except one or two cleanup in the Consensus class that almost nobody use it should be fine.

**Comment by dangershony** [comment]
>A constructor of BitcoinStream should be added with Network as parameter, and the current constructor should be marked [Obsolete].

This sounds very good, I already started such a PR but got distracted.

A `BlockHeaderFactory` also sounds like a good start and later we might add more factories for Block and Transaction

Currently in our code we use static flags to determine serialization but this has to go away, we have a flag in Network that specifies if this is a PoS network so this covers cases for serialization (for example we with block we need to add a block signature and in transaction the PoS timestamp)

But ideally for the stratis platform we need to be able to also use different Block/Header/Transaction types as in the future some of our altcoins might add more data to a trx (or not use a UTXO style trx all) but we cross this bridge when we get there.

**Comment by NicolasDorier** [comment]
@dangershony actually I am thinking: `BitcoinStream` does not have to know the `Network`. The reason is that under the model of subclassing BlockHeader/Transaction/Block, those instances knows how to serialize themselves so `BitcoinStream` does not need to know.

**Comment by NicolasDorier** [comment]
So actually I think the plan might be quite easy:
Really the only thing those factory need is a `.CreateBlockHeader()` method.
We then only have to deprecate people to use `new BlockHeader()`.

**Comment by Kukks** [comment]
Will give this a go this Monday. :)

**Comment by dangershony** [comment]
@NicolasDorier you are right but having the `Network` in `BitcoinStream` may still be very useful for all kinds of scenarios, I would still try to get it in there if we can. what you suggested will work without breaking anything.

**Comment by NicolasDorier** [comment]
@dangershony this is too difficult I think. Methods like GetSerializedSize, GetVirtualSize does not expect you pass the Network to them, and should not be expecting this either.

**Comment by NicolasDorier** [comment]
@Kukks try to keep GetPowHash if you can, else dump it. It might break stuff but this is not a very used piece of code and was not well designed.

**Comment by Kukks** [comment]
@NicolasDorier  

I can mark the `GetPoWHash` in Network as `Obsolete` to raise compile warnings on it.  I can't see how leaving the GetPoWHash setter would work though. 

Are there any usecases of anyone using the GetPoWHash other than in the BlockHeader?

**Comment by NicolasDorier** [comment]
Nah, it should be in the block header. That's why maybe removing it is better.

**Comment by Kukks** [comment]
@NicolasDorier 
```
The GetPoWHash of BlockHeader currently takes a Network parameter. This is not needed anymore as the BlockHeader would get the logic for GetPowHash calculation inside itself.
We can just keep the current GetPowHash(Network) but mark it [Obsolete].
Then having the real GetPowHash() without Network parameter.
```

I couldn't find a signature for ` GetPowHash(Network) ` in BlockHeader  or anywhere else in the project? 
I will be getting rid of the GetPowHash in Network as it seems messy to try and handle that config when the block header provided from a factory should be handling it.

**Comment by NicolasDorier** [comment]
@Kukks yes `GetPowHash()` should be in header, it was not until now.

**Comment by Kukks** [comment]
@NicolasDorier I see there has been a number of changes pushed. It seems all the types needed to be covered for STRAT (@dangershony ) are covered by the `ConsensusFactory`.

**Comment by Kukks** [comment]
Maybe the Altcoins integrated should be in their own nuget and not directly in NBitcoin?

**Comment by NicolasDorier** [comment]
Yes. I worked quite a lot today on this. Now NBitcoin has `NBitcoin.TestFramework` and `NBitcoin.Altcoins` in this repo. Currently implemented is Litecoin, BCH, Feathercoin. I would like to try with an alt having different shape.

I updated NBXplorer and BTCPay to use this new version.
